src/System.Management.Automation/engine/SessionState.cs

```diff
@@ -487,34 +482,6 @@ private SessionStateEntryVisibility checkPathVisibility(List<string> list, strin
 
         #endregion Private data
 
-        /// <summary>
-        /// Notification for SessionState to do cleanup
-        /// before runspace is closed.
-        /// </summary>
-        ///
-        internal void RunspaceClosingNotification()
-        {
-            if (this != ExecutionContext.TopLevelSessionState && Providers.Count > 0)
-            {
-                // Remove all providers at the top level...
-
-                CmdletProviderContext context = new CmdletProviderContext(this.ExecutionContext);
-
-                Collection<string> keys = new Collection<string>();
-                foreach (string key in Providers.Keys)
```

Comment:
I'm not sure we should remove this method, but this extra loop is pointless - just copying the keys to a temp collection.

Reply:
I wasn't sure about this method.  I set a breakpoint in it and couldn't get it to hit within the if statement using runspace creation.

Reply:
The if statement looks to me is checking if the current `EngineSessionState` is a module session state and if the module defines providers. This seems will happen when a Runspace is closing while running cmdlets from a module.

`RemoveProvider(string providerName, bool force, CmdletProviderContext context)` doesn't look like solely related to `RunspaceConfigration`.

I think we should keep this method but remove the extra loop.